clang/lib/Sema/CheckedCAnalysesPrepass.cpp

```diff
@@ -80,20 +93,50 @@ class PrepassHelper : public RecursiveASTVisitor<PrepassHelper> {
 
       // We add VarWithBounds to the set of all variables in whose bounds
       // expressions V occurs.
-      if (VarWithBounds) {
-        auto It = Info.BoundsVars.find(V);
-        if (It != Info.BoundsVars.end())
+      if (!VarWithBounds)
+        return true;
+
+      if (InBoundsExprLower) {
+        auto It = Info.BoundsVarsLower.find(V);
+        if (It != Info.BoundsVarsLower.end())
+          It->second.insert(VarWithBounds);
+        else {
+          VarSetTy Vars;
+          Vars.insert(VarWithBounds);
+          Info.BoundsVarsLower[V] = Vars;
+        }
+      } else if (InBoundsExprUpper) {
+        auto It = Info.BoundsVarsUpper.find(V);
+        if (It != Info.BoundsVarsUpper.end())
           It->second.insert(VarWithBounds);
         else {
           VarSetTy Vars;
           Vars.insert(VarWithBounds);
-          Info.BoundsVars[V] = Vars;
+          Info.BoundsVarsUpper[V] = Vars;
         }
       }
 
       return true;
     }
 
+    bool VisitBoundsExpr(BoundsExpr *B) {
+      if (!VarWithBounds)
+        return true;
+
+      BoundsExpr *Bounds = SemaRef.ExpandBoundsToRange(VarWithBounds, B);
```

Comment:
`SemaRef.NormalizeBounds` will attach the expanded range bounds to `VarWithBounds` to avoid making repeated calls to `ExpandBoundsToRange` on `VarWithBounds` in the future.